### PR TITLE
[SCIP] Ipopt integration and bliss patch

### PR DESCRIPTION
**Pull request by matbesancon**
This PR adds Ipopt to the dependencies, letting SCIP use it for non-linear problems.

It also uses the [bliss patch](https://github.com/ds4dm) from @AntoinePrv as recommended for SCIP.
I didn't incorporate the bliss patch currently on Yggdrasil as we still consider windows compilation as broken and low priority for SCIP_jll for now.

Trying to build locally with debug, still two issues:
- Cmake does not seem to find bliss, should it be installed in another location?
- Cmake does not find GMP, which is more problematic. It didn't happen before I believe. Also, the current master of BB does not accept GMP v0.6.1 which is the recommended one in this build

### CODE REVIEW

**Comment by matbesancon** [comment]
The cmake error in the log is new:
```
[10:19:14] CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
[10:19:14]    RUN_RESULT (advanced)
[10:19:14] For details see /workspace/srcdir/scipoptsuite-7.0.2/build/TryRunResults.cmake
```

Here:
https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=8369&view=logs&j=110ff2ac-d786-5f9d-ad4a-4416198d34bc&t=d1b307f4-02aa-555e-2ce3-c5224e6e2367&l=6274

File: S/SCIP/build_tarballs.jl [hunk 1]
```diff
@@ -7,16 +7,23 @@ version = v"0.1.2"
...
 ]
 
 # Bash recipe for building across all platforms
 script = raw"""
 cd $WORKSPACE/srcdir
-cd scipoptsuite*
+cd bliss
+mkdir build
+cd build
+cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_BUILD_TYPE=Release ..
```
Comment (fingolfin) [inline S/SCIP/build_tarballs.jl:20]: You are not cross compiling bliss. At the very least you'll need something like `-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}` below. Perhaps take a look at the Bliss

**Comment by fingolfin** [comment]
There is no GMP v0.6.1 , not sure what you mean there?

**Comment by matbesancon** [comment]
> There is no GMP v0.6.1 , not sure what you mean there?

sorry meant 6.1

**Comment by fingolfin** [comment]
BB accepts GMP 6.1.2 just fine; however, the current Julia master and release-1.6 branches do not. I think @giordano has an issue open for that somewhere (couldn't find it right now); basically you need a Julia from before the "fake JLL" work was merged. That happened in Julia commit 2eccd8e086ba1b08dd8b4d04e86b548fcf8ec5e5, so any commit before that should do it.

**Comment by matbesancon** [comment]
> BB accepts GMP 6.1.2 just fine; however, the current Julia master and release-1.6 branches do not 

The latest BB, which is the one running on PRs I believe, only runs on Julia 1.6+

**Comment by matbesancon** [comment]
The Julia version I have locally does not change the problem that the PR would fail would it?

**Comment by giordano** [comment]
> The Julia version I have locally does not change the problem that the PR would fail would it?

As Max said, the version of Julia we run here is happy to accept a different version of GMP than the one which comes with Julia v1.6

**Comment by matbesancon** [comment]
OK so I'll leave it as-is, the version constraint was maintained.

**Comment by fingolfin** [comment]
The error which fails this seems to be the following:
```
[21:02:06] CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
[21:02:06]    RUN_RESULT (advanced)
[21:02:06] For details see /workspace/srcdir/scipoptsuite-7.0.2/build/TryRunResults.cmake
```
That would suggest a cross compilation issue in the SCIP build system... which is a bit odd, though, as presumably this used to work with the exact same SCIP version?

My suggestion for debugging this would be to not try to much at once; perhaps first edit the PR to basically be unchanged (i.e., comment out part building bliss, revert the final cmake invocation to what it was); then see if that version of the PR builds (you may want to convert it to a draft PR first, to signal that it shouldn't be merged in that state). If that still breaks, we need to see what changed since the last time it was built successfully. If it works, you can next try removing the `-DIPOPT=0` again; and separately try switching the bliss. That should narrow down where it goes wrong (right now my suspicion is on the `-DIPOPT=0` removal -- since I am guessing that removal caused code to be run/built that wasn't before... but I am just speculating here)

**Comment by matbesancon** [comment]
Thanks, I started reverting some changes and see where it passes

**Comment by matbesancon** [comment]
Ok all good, now re-adding Ipopt

**Comment by matbesancon** [comment]
Ok so it seems that even with `-DIPOPT=0`, SCIP looks for it?
```
[ Info: Found a valid dl path libipopt.so while looking for libscip
[ Info: Found a valid dl path libipopt.so.3 while looking for libscip
[ Info: Found a valid dl path libipopt.so.3.13.2 while looking for libscip
```

https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=8468&view=logs&j=905ed0a7-02b1-5a23-336c-a2ad3950fe27&t=b86956c0-701d-576e-a756-1965f354c711&l=8510

**Comment by matbesancon** [comment]
Ok no all good, Ipopt was ignored:
```
2021-01-17T12:44:42.2981754Z [0m[1m[12:44:42] [22m-- The following OPTIONAL packages have been found:
2021-01-17T12:44:42.2982992Z [0m[1m[12:44:42] [22m
2021-01-17T12:44:42.2983894Z [0m[1m[12:44:42] [22m * Quadmath
2021-01-17T12:44:42.2984711Z [0m[1m[12:44:42] [22m * GMP
2021-01-17T12:44:42.2985392Z [0m[1m[12:44:42] [22m * BLISS
2021-01-17T12:44:42.2986053Z [0m[1m[12:44:42] [22m
2021-01-17T12:44:42.2986958Z [0m[1m[12:44:42] [22m-- The following REQUIRED packages have been found:
2021-01-17T12:44:42.2987778Z [0m[1m[12:44:42] [22m
2021-01-17T12:44:42.2988477Z [0m[1m[12:44:42] [22m * SOPLEX
2021-01-17T12:44:42.2989156Z [0m[1m[12:44:42] [22m * SCIP
2021-01-17T12:44:42.2989793Z [0m[1m[12:44:42] [22m
2021-01-17T12:44:42.2990769Z [0m[1m[12:44:42] [22m-- The following OPTIONAL packages have not been found:
2021-01-17T12:44:42.2991608Z [0m[1m[12:44:42] [22m
2021-01-17T12:44:42.2992253Z [0m[1m[12:44:42] [22m * ZLIB
2021-01-17T12:44:42.2993281Z [0m[1m[12:44:42] [22m * CRITERION
```

**Comment by matbesancon** [comment]
OK Ipopt integration was successful:
```
```

**Comment by matbesancon** [comment]
OK now on to adding SYM=bliss. This shouldn't change a thing since bliss is already detected

**Comment by matbesancon** [comment]
still good, removing enforced GMP version

File: S/SCIP/build_tarballs.jl [hunk 2]
```diff
@@ -49,8 +62,9 @@ products = [
 # Dependencies that must be installed before this package can be built
 dependencies = [
     Dependency(PackageSpec(name="bliss_jll", uuid="508c9074-7a14-5c94-9582-3d4bc1871065")),
-    Dependency(PackageSpec(name="GMP_jll", uuid="781609d7-10c4-51f6-84f2-b8444358ff6d"), v"6.1.2"),
+    Dependency(PackageSpec(name="GMP_jll", uuid="781609d7-10c4-51f6-84f2-b8444358ff6d")),
```
Comment (fingolfin) [inline S/SCIP/build_tarballs.jl:65]: This is wrong unless you want this JLL to be only usable in Julia >= 1.6 (and then you really should add a corresponding julia_compat value)

Comment (matbesancon) [reply S/SCIP/build_tarballs.jl:65]: got it thank you, reverted

**Comment by matbesancon** [comment]
So we now have a compiling SCIP + Ipopt. I just removed the bliss repo for now to avoid useless network delays

**Comment by matbesancon** [comment]
@rschwarz we'll have Ipopt in the JLL after this one for the PR :)

**Comment by matbesancon** [comment]
@fingolfin thanks for the help along the rocky road. The PR is now at a satisfactory point, the bliss fork does not seem ready for use now it will need further investigation and will come in another PR

**Pull request merged**